File: packages/contracts/ts/v1/order-ack.ts
import type { OrderStatus } from './enums';

/**
 * Rate limit metadata included in rejection responses.
 */
export interface RateLimitInfo {
  /** Rate limit scope: "user", "contest", or "global" */
  scope: 'user' | 'contest' | 'global';
  /** Maximum requests allowed in the window */
  limit: number;
  /** Window duration (e.g., "1s", "1m") */
  window: string;
  /** Milliseconds until next request allowed */
  retry_after_ms: number;
}

/**
 * Acknowledgment response for an order request.
 */
export interface OrderAck {
  order_id: string;
  status: OrderStatus;
  reason?: string;
  /** Present only for RATE_LIMITED rejections */
  rate_limit?: RateLimitInfo;
}

File: packages/contracts/v1/order_ack.go
package v1

// RateLimitInfo contains metadata about a rate limit rejection.
type RateLimitInfo struct {
	Scope        string `json:"scope"`          // "user", "contest", or "global"
	Limit        int    `json:"limit"`          // Maximum requests allowed
	Window       string `json:"window"`         // Window duration (e.g., "1s", "1m")
	RetryAfterMs int64  `json:"retry_after_ms"` // Milliseconds until next request allowed
}

// OrderAck represents an acknowledgment response for an order request.
type OrderAck struct {
	OrderID   string         `json:"order_id"`
	Status    OrderStatus    `json:"status"`
	Reason    *string        `json:"reason,omitempty"`
	RateLimit *RateLimitInfo `json:"rate_limit,omitempty"` // Present only for RATE_LIMITED rejections
}
